atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/ConcludedTransactionsUpdaterTask.java

```diff
@@ -0,0 +1,96 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.sweep;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Range;
+import com.palantir.atlasdb.AtlasDbConstants;
+import com.palantir.atlasdb.sweep.queue.ShardAndStrategy;
+import com.palantir.atlasdb.sweep.queue.ShardProgress;
+import com.palantir.atlasdb.table.description.SweepStrategy.SweeperStrategy;
+import com.palantir.atlasdb.transaction.knowledge.CoordinationAwareKnownConcludedTransactionsStore;
+import java.util.Comparator;
+import java.util.Set;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public final class ConcludedTransactionsUpdaterTask implements AutoCloseable {
+    private final Set<ShardAndStrategy> allShardsAndStrategies;
+    private final ShardProgress progress;
+    private final CoordinationAwareKnownConcludedTransactionsStore concludedTransactionsStore;
+    private final ScheduledExecutorService executor;
+
+    private ConcludedTransactionsUpdaterTask(
+            Set<ShardAndStrategy> allShardsAndStrategies,
+            CoordinationAwareKnownConcludedTransactionsStore concludedTransactionsStore,
+            ShardProgress progress,
+            ScheduledExecutorService executorService) {
+        this.allShardsAndStrategies = allShardsAndStrategies;
+        this.concludedTransactionsStore = concludedTransactionsStore;
+        this.progress = progress;
+        this.executor = executorService;
+    }
+
+    public static ConcludedTransactionsUpdaterTask create(
+            int numShards,
+            CoordinationAwareKnownConcludedTransactionsStore concludedTransactionsStore,
+            ShardProgress progress,
+            ScheduledExecutorService executor) {
+        ConcludedTransactionsUpdaterTask task = new ConcludedTransactionsUpdaterTask(
+                getAllShardsAndStrategies(numShards), concludedTransactionsStore, progress, executor);
+        task.schedule();
+        return task;
+    }
+
+    private void schedule() {
+        executor.scheduleWithFixedDelay(
+                this::runOneIteration,
+                AtlasDbConstants.CONCLUDED_TRANSACTIONS_UPDATE_INITIAL_DELAY_MILLIS,
+                AtlasDbConstants.CONCLUDED_TRANSACTIONS_UPDATE_TASK_DELAY_MILLIS,
+                TimeUnit.MILLISECONDS);
+    }
+
+    private void runOneIteration() {
+        long minLastSweptTimestamp = allShardsAndStrategies.stream()
+                .map(progress::getLastSweptTimestamp)
+                .min(Comparator.naturalOrder())
+                .orElse(1L);
+
+        Range<Long> concludedTimestamps = Range.closed(1L, minLastSweptTimestamp);
```

Comment:
should check with @gmaretic: I think we need 0L. 1L is an actual timestamp, and so it's possible that something was written starting at timestamp 1 so we don't necessarily want to assume it's concluded.

Reply:
my bad i read SweepQueueUtils.INITIAL_TIMESTAMP as 1 instead of -1;